src/solvers/refinement/string_refinement.cpp

```diff
@@ -0,0 +1,869 @@
+    }
+    else
+    {
+      add_lemma(generator.axioms[i]);
+    }
+
+  initial_index_set(universal_axioms);
+  update_index_set(cur);
+  cur.clear();
+  add_instantiations();
+
+  while((initial_loop_bound--)>0)
+  {
+    decision_proceduret::resultt res=supert::dec_solve();
+
+    switch(res)
+    {
+    case D_SATISFIABLE:
+      if(!check_axioms())
+      {
+        debug() << "check_SAT: got SAT but the model is not correct" << eom;
+      }
+      else
+      {
+        debug() << "check_SAT: the model is correct" << eom;
+        return D_SATISFIABLE;
+      }
+
+      debug() <<  "refining..." << eom;
+      // Since the model is not correct although we got SAT, we need to refine
+      // the property we are checking by adding more indices to the index set,
+      // and instantiating universal formulas with this indices.
+      // We will then relaunch the solver with these added lemmas.
+      current_index_set.clear();
+      update_index_set(cur);
+      cur.clear();
+      add_instantiations();
+
+      if(variable_with_multiple_occurence_in_index)
+        // This can happen depending on the universal properties that were
+        // given to the solver, however we can give a guarantee about the
+        // termination of the solver.
+        debug() << "WARNING: some variable appears multiple times" << eom;
+
+      if(current_index_set.empty())
+      {
+        debug() << "current index set is empty" << eom;
+        return D_SATISFIABLE;
+      }
+
+      display_index_set();
+      debug()<< "instantiating NOT_CONTAINS constraints" << eom;
+      for(unsigned i=0; i<not_contains_axioms.size(); i++)
+      {
+        debug()<< "constraint " << i << eom;
+        std::vector<exprt> lemmas;
+        instantiate_not_contains(not_contains_axioms[i], lemmas);
+        for(unsigned j=0; j<lemmas.size(); j++)
+          add_lemma(lemmas[j]);
+      }
+      break;
+    default:
+      return res;
+    }
+  }
+  debug() << "string_refinementt::dec_solve reached the maximum number"
+           << "of steps allowed" << eom;
+  return D_ERROR;
+}
+
+
+bvt string_refinementt::convert_bool_bv(const exprt &boole, const exprt &orig)
+{
+  bvt ret;
+  ret.push_back(convert(boole));
+  size_t width=boolbv_width(orig.type());
+  for(size_t i=1; i<width; i++)
+    ret.push_back(const_literal(false));
+  return ret;
+}
+
+void string_refinementt::add_lemma(const exprt &lemma, bool add_to_index_set)
+{
+  if(!seen_instances.insert(lemma).second)
+    return;
+
+  if(lemma==true_exprt())
+  {
+    debug() << "string_refinementt::add_lemma : tautology" << eom;
+    return;
+  }
+
+  debug() << "adding lemma " << from_expr(lemma) << eom;
+
+  prop.l_set_to_true(convert(lemma));
+  if(add_to_index_set)
+    cur.push_back(lemma);
+}
+
+unsigned to_unsigned_integer(const constant_exprt & expr)
+{
+  mp_integer i;
+  if(to_integer(expr, i))
+    return 0;
+  if(i<0)
+    return 0;
+  else
+    return integer2unsigned(i);
+}
```

Comment:
This looks like a general helper function that should go in arith_tools.h rather than hiding in here.